File: assignments/haskell/zipper/Zipper.hs
module Zipper (
    BinTree(..),
    Zipper,

    fromTree,
    toTree,

    value,
    left,
    right,
    up,

    setValue,
    setLeft,
    setRight
) where

-- | A binary tree.
data BinTree a = BT { 
    btValue :: a                 -- ^ Value
  , btLeft  :: Maybe (BinTree a) -- ^ Left child
  , btRight :: Maybe (BinTree a) -- ^ Right child
} deriving (Eq, Show)

-- | A zipper for a binary tree.
data Zipper a -- Complete this definition

-- | Get a zipper focussed on the root node.
fromTree :: BinTree a -> Zipper a
fromTree = undefined

-- | Get the complete tree from a zipper.
toTree :: Zipper a -> BinTree a
toTree = undefined

-- | Get the value of the focus node.
value :: Zipper a -> a
value = undefined 

-- | Get the left child of the focus node, if any.
left :: Zipper a -> Maybe (Zipper a)
left = undefined

-- | Get the right child of the focus node, if any.
right :: Zipper a -> Maybe (Zipper a)
right = undefined

-- | Get the parent of the focus node.
--
-- If the focus node is the root node doesn't move the zipper.
up :: Zipper a -> Zipper a
up = undefined

-- | Set the value of the focus node.
setValue :: a -> Zipper a -> Zipper a
setValue = undefined

-- | Replace a left child tree.
setLeft :: Maybe (BinTree a) -> Zipper a -> Zipper a
setLeft = undefined 

-- | Replace a right child tree.
setRight :: Maybe (BinTree a) -> Zipper a -> Zipper a
setRight = undefined

File: assignments/elixir/zipper/zipper.exs
defrecord BinTree, value: nil, left: nil, right: nil do
  @moduledoc """
  A node in a binary tree.

  `value` is the value of a node.
  `left` is the left subtree (nil if no subtree).
  `right` is the right subtree (nil if no subtree).
  """
  record_type value: any, left: BinTree.t | nil, right: BinTree.t | nil
end

defmodule Zipper do
  @doc """
  Get a zipper focussed on the root node.
  """
  @spec from_tree(BT.t) :: Z.t
  def from_tree(bt)

  @doc """
  Get the complete tree from a zipper.
  """
  @spec to_tree(Z.t) :: BT.t
  def to_tree(z)

  @doc """
  Get the value of the focus node.
  """
  @spec value(Z.t) :: any
  def value(z)

  @doc """
  Get the left child of the focus node, if any.
  """
  @spec left(Z.t) :: Z.t | nil
  def left(z)
  
  @doc """
  Get the right child of the focus node, if any.
  """
  @spec right(Z.t) :: Z.t | nil
  def right(z)

  @doc """
  Get the parent of the focus node, if any.
  """
  @spec up(Z.t) :: Z.t
  def up(z)

  @doc """
  Set the value of the focus node.
  """
  @spec set_value(Z.t, any) :: Z.t
  def set_value(z, v)
  
  @doc """
  Replace the left child tree of the focus node. 
  """
  @spec set_left(Z.t, BT.t) :: Z.t
  def set_left(z, l)
  
  @doc """
  Replace the right child tree of the focus node. 
  """
  @spec set_right(Z.t, BT.t) :: Z.t
  def set_right(z, r)
end
